synapse/storage/databases/state/bg_updates.py

```diff
@@ -182,69 +212,35 @@ def _get_state_groups_from_groups_txn(
                     ) {where_clause}
                     ORDER BY type, state_key, state_group DESC
                 """
+            else:
+                # SQLite doesn't support `SELECT DISTINCT ON`, so we have to just get
+                # some potential duplicate (type, state_key) pairs and then only use the
+                # first of each kind we see.
+                overall_select_clause = f"""
+                    SELECT type, state_key, event_id
+                    FROM state_groups_state
+                    WHERE state_group IN (
+                        SELECT state_group FROM sgs
+                    ) {where_clause}
+                    ORDER BY type, state_key, state_group DESC
+                """
```

Comment:
This query is exactly the same as the Postgres version above except for the `DISTINCT ON (type, state_key)` part at the beginning. Any interest in sharing the bulk of the query here? Not sure it would be more clear